spanconfig/sqlwatcher: hide factory pattern, speed up tests, surface no-op checkpoints periodically

irfansharif opened the pull request:
See individual commits.

---

**spanconfig/sqlwatcher: speed up tests**

Using a more aggressive closed timestamp target duration brings the runtime for
TestSQLWatcherReactsToUpdate down from 45s+ to single digits. To speed it up,
we also just re-use the same SQLWatcher instead of creating a new one for every
test case.

Other tests in the package benefit from a similar treatment. Using the default
target duration in fact masked a buggy test; in a future commit we end up
rewrite that test so it's skipped for now.

**spanconfig/sqlwatcher: hide the factory pattern**

The only mutable state across concurrent WatchForSQLUpdate calls was the
internal buffer, which does not need to hang off the surrounding struct.  This
obviates the need for the factory pattern we were using -- callers can set up
multiple SQLWatchers concurrently as is (see TestSQLWatcherMultiple).

This PR first hides the factory under the package boundary; in a later commit
it shed it altogether. This has the benefit of reducing the number of symbols
in pkg/spanconfig and making it symmetric with the other spanconfig
dependencies typically found together (KVAccessor, SQLTranslator). It's also
every-so-slightly less verbose to use in the upcoming spanconfig.Reconciler
(#71994).

**spanconfig/sqlwatcher: checkpoint, even without updates**

If `system.{descriptor,zones}` is static, we would previously discard
all checkpoint events. This is not what we want -- we'd still like to
know how caught up we are, with ongoing updates or otherwise.

MVCC GC, after all, can still happen; if our last checkpoint was when
the last update occurred, we may end up doing a lot of unnecessary work
when finding out our last checkpoint was GC-ed from underneath us. Lack
of periodic checkpoints also makes tests difficult to write -- you'd
have to induce a benign update to flush out all earlier ones. This
latent flakiness was uncovered after speeding up some existing tests in
an earlier commit.

NB: For incremental (possibly noop) checkpoints, we need to ensure that
the sqlwatcher's buffer is initialized with a low watermark. When
flushing events, we take the most conservative timestamp. If not
initialized to a high water, this might be 0 -- violating the
sqlwatcher's monotonically increasing checkpoint ts invariant.

Release note: None

irfansharif committed:
spanconfig/sqlwatcher: hide the factory pattern

The only mutable state across concurrent WatchForSQLUpdate calls was the
internal buffer, which does not need to hang off the surrounding struct.
This obviates the need for the factory pattern we were using -- callers
can set up multiple SQLWatchers concurrently as is (see
TestSQLWatcherMultiple).

This commit simply hides the factory under the package boundary; in a
future commit we'll shed it altogether. This has the benefit of reducing
the number of symbols in pkg/spanconfig and making it symmetric with the
other spanconfig dependencies typically found together (KVAccessor,
SQLTranslator). It's also every-so-slightly less verbose to use in the
upcoming spanconfig.Reconciler (#71994).

Release note: None

irfansharif committed:
spanconfig/sqlwatcher: speed up tests

Using a more aggressive closed timestamp target duration brings the
runtime for TestSQLWatcherReactsToUpdate down from 45s+ to single
digits. To speed it up, we also just re-use the same sqlwatcher instead
of creating a new one for every test case.

Other tests in the package benefit from a similar treatment. Using the
default target duration in fact masked a buggy test; in a future commit
we end up rewrite that test so it's skipped for now.

Release note: None

irfansharif committed:
spanconfig/sqlwatcher: remove internal factory pattern

A previous commit obviated the need for a factory altogether, but went
only as far hiding it within the package boundary. This commit gets rid
of it entirely.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73171)
<!-- Reviewable:end -->

irfansharif committed:
spanconfig/sqlwatcher: checkpoint, even without updates

If `system.{descriptor,zones}` is static, we would previously discard
all checkpoint events. This is not what we want -- we'd still like to
know how caught up we are, with ongoing updates or otherwise.

MVCC GC, after all, can still happen; if our last checkpoint was when
the last update occurred, we may end up doing a lot of unnecessary work
when finding out our last checkpoint was GC-ed from underneath us. Lack
of periodic checkpoints also makes tests difficult to write -- you'd
have to induce a benign update to flush out all earlier ones. This
latent flakiness was uncovered after speeding up some existing tests in
an earlier commit.

NB: For incremental (possibly noop) checkpoints, we need to ensure that
the sqlwatcher's buffer is initialized with a low watermark. When
flushing events, we take the most conservative timestamp. If not
initialized to a high water, this might be 0 -- violating the
sqlwatcher's monotonically increasing checkpoint ts invariant.

Release note: None

nvb reviewed (approved):
> Reviewed 2 of 2 files at r1, 10 of 10 files at r2, 1 of 1 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73171)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @irfansharif)

___
*[pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go, line 165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73171#-Mpixz5R9G1RWFq0rB8V:-Mpixz5R9G1RWFq0rB8W:b-ogoyzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ae233c5f6e8854bad0eb8522dd019b3b350334/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go#L165)):*
> ```Go
> 				return err
> 			}
> 			if len(events) == 0 && !checkpointNoops.ShouldProcess(timeutil.Now()) {
> ```

Would it be straightforward to write a test for this?
___
*[pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher_test.go, line 371 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73171#-MpivH765bZ22mFq_0zC:-MpivH765bZ22mFq_0zD:btxylbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee68771412fb4a6a2e3fd34c77071d9a32cbe423/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher_test.go#L371)):*
> ```Go
> 				return
> 			default:
> 				time.Sleep(100 * time.Millisecond)
> ```

Why was this needed?
___
*[pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher_test.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73171#-MpiwawU0uxshXJqUxhA:-MpiwawU0uxshXJqUxhB:b4n1836) ([raw file](https://github.com/cockroachdb/cockroach/blob/46c7597875cc01414aa0384c2e5f1e299961211a/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher_test.go#L320)):*
> ```Go
> 	}
> 
> 	wg.Wait()
> ```

nit: consider pulling this call into the branch that includes the goroutines it's waiting on.


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
Thanks Nathan! Merging on green.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73171#-:-MpktkCQBIgM_foJbGgD:bpr7sxo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go, line 165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73171#-Mpixz5R9G1RWFq0rB8V:-MpktVCv76ZXnXfRIf1C:b-y8vepy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ae233c5f6e8854bad0eb8522dd019b3b350334/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go#L165)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Would it be straightforward to write a test for this?
</blockquote></details>

It would be, done.
___
*[pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher_test.go, line 371 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73171#-MpivH765bZ22mFq_0zC:-MpktaRWDlVacSKKEVLi:bbs9wt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee68771412fb4a6a2e3fd34c77071d9a32cbe423/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher_test.go#L371)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why was this needed?
</blockquote></details>

We were busy looping before. The wait isn't "needed" for anything, but I don't think busy looping is what we wanted here either.


<!-- Sent from Reviewable.io -->

irfansharif commented:
bors r+
